src/Lookup/EntityRetrievingClosestReferencedEntityIdLookup.php

```diff
@@ -0,0 +1,236 @@
+		$this->maxDepth = $maxDepth;
+		$this->maxEntityVisits = $maxEntityVisits;
+	}
+
+	/**
+	 * Get the closest entity (out of $toIds), from a given entity. The starting entity, and
+	 * the target entities are (potentially indirectly, via intermediate entities) linked by
+	 * statements with the given property ID, pointing from the starting entity to one of the
+	 * target entities.
+	 *
+	 * @since 3.10
+	 *
+	 * @param EntityId $fromId
+	 * @param PropertyId $propertyId
+	 * @param EntityId[] $toIds
+	 *
+	 * @return EntityId|null Returns null in case none of the target entities are referenced.
+	 * @throws ReferencedEntityIdLookupException
+	 */
+	public function getReferencedEntityId( EntityId $fromId, PropertyId $propertyId, array $toIds ) {
+		if ( !$toIds ) {
+			return null;
+		}
+
+		$this->alreadyVisited = [];
+
+		$steps = $this->maxDepth + 1; // Add one as checking $fromId already is a step
+		$toVisit = [ $fromId ];
+
+		while ( $steps-- ) {
+			$this->entityPrefetcher->prefetch( $toVisit );
+			$toVisitNext = [];
+
+			foreach ( $toVisit as $curId ) {
+				$result = $this->processEntityById( $curId, $fromId, $propertyId, $toIds, $toVisitNext );
+				if ( $result ) {
+					return $result;
+				}
+			}
+			// Remove already visitted entities
+			$toVisit = array_unique(
+				array_diff( $toVisitNext, array_keys( $this->alreadyVisited ) )
+			);
+
+			if ( !$toVisit ) {
+				return null;
+			}
+		}
+
+		// Exhausted the max. depth without finding anything.
+		throw new MaxReferenceDepthExhaustedException(
+			$fromId,
+			$propertyId,
+			$toIds,
+			$this->maxDepth
+		);
+	}
+
+	/**
+	 * Find out whether an entity (directly) references one of the target ids.
+	 *
+	 * @param EntityId $id Id of the entity to process
+	 * @param EntityId $fromId Id this lookup started from
+	 * @param PropertyId $propertyId
+	 * @param EntityId[] $toIds
+	 * @param EntityId[] &$toVisit List of entities that still need to be checked
+	 * @return EntityId|null Target id the entity refers to, null if none.
+	 */
+	private function processEntityById(
+		EntityId $id,
+		EntityId $fromId,
+		PropertyId $propertyId,
+		array $toIds,
+		array &$toVisit
+	) {
+		$entity = $this->getEntity( $id, $fromId, $propertyId, $toIds );
+		if ( !$entity ) {
+			return null;
+		}
+
+		$mainSnaks = $this->getMainSnaks( $entity, $propertyId );
+
+		foreach ( $mainSnaks as $mainSnak ) {
+			$result = $this->processSnak( $mainSnak, $toVisit, $toIds );
+			if ( $result ) {
+				return $result;
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * @param EntityId $id Id of the entity to get
+	 * @param EntityId $fromId Id this lookup started from
+	 * @param PropertyId $propertyId
+	 * @param EntityId[] $toIds
+	 *
+	 * @return StatementListProvider|null Null if not applicable.
+	 */
+	private function getEntity( EntityId $id, EntityId $fromId, PropertyId $propertyId, array $toIds ) {
+		$this->alreadyVisited[$id->getSerialization()] = true;
```

Comment:
I miss a safety check here that starts screaming when this is called with an already visited entity. As far as I can see this is not required, as the caller makes sure this can't happen. But this is a little hidden in an array_diff, and I would feel better with an additional isset check. It's cheap.

Reply:
Added a `LogicException` here.

Reply:
Changed this to actually trigger a warning now… throwing an exception is maybe a bit harsh.

Reply:
`trigger_error` causes a global side effect. Without knowing the context here I can't say what is appropriate, though normally you either throw an exception and let the caller deal with it, which might be ignoring the error, escalating it, logging it, etc. Or, in cases where the execution in the local context should not be aborted, you log something or otherwise output info via some collaborator (which could also be an event thing or whatever).